src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -752,6 +730,116 @@ public Integer updateAppsUsageAssociationData(
         return ImmutableList.of();
     }
 
+    private String getFirstValueOrNullByUUID( SetMultimap<UUID, Object> entity, FullQualifiedName fqn ) {
+        UUID fqnId = propertyTypeIdsByFQN.get( fqn );
+        Object value = Iterables.getFirst( entity.get( fqnId ), null );
+
+        return value == null ? null : value.toString();
+    }
+
+    private Set<OffsetDateTime> getDateTimeValuesFromDeviceData( List<SetMultimap<UUID, Object>> data ) {
+        Set<OffsetDateTime> dateTimes = new HashSet<>();
+        data.forEach(
+                entity -> {
+                    // most date properties in the entity are of length 1
+                    for ( Object date : entity.get( propertyTypeIdsByFQN.get( DATE_LOGGED_FQN ) ) ) {
+                        OffsetDateTime parsedDateTime = OffsetDateTime
+                                .parse( date.toString() );
+
+                        // filter out problematic entities with dates in the sixties
+                        if ( parsedDateTime.isAfter( MINIMUM_DATE ) ) {
+                            dateTimes.add( parsedDateTime );
+                        }
+                    }
+                }
+        );
+        return dateTimes;
+    }
+
+    private Set<EntityKey> getAllEntityKeys(
+            List<SetMultimap<UUID, Object>> data,
+            UUID participantEKID,
+            String participantId,
+            String deviceId ) {
+        Set<EntityKey> entityKeys = new LinkedHashSet<>();
+        Map<UUID, Set<Object>> entityData = new HashMap<>();
+
+        // step 1: create entity keys used in createUserAppsEntitiesAndAssociations fn
+        for ( SetMultimap<UUID, Object> entity : data ) {
+            String appPackageName, appName;
+            appPackageName = appName = getFirstValueOrNullByUUID( entity, FULL_NAME_FQN );
+
+            if ( entity.containsKey( propertyTypeIdsByFQN.get( TITLE_FQN ) ) ) {
+                appName = getFirstValueOrNullByUUID( entity, TITLE_FQN );
+            }
+
+            // user apps entity key
+            entityData.put( propertyTypeIdsByFQN.get( FULL_NAME_FQN ), Sets.newHashSet( appPackageName ) );
+            entityData.put( propertyTypeIdsByFQN.get( TITLE_FQN ), Sets.newHashSet( appName ) );
+
+            EntityKey userAppEK = getUserAppsEntityKey( entityData );
+            entityKeys.add( userAppEK );
+
+            // recordedby entity key
+            String dateLogged = getMidnightDateTime( getFirstValueOrNullByUUID( entity, DATE_LOGGED_FQN ) );
+            entityData.put( propertyTypeIdsByFQN.get( DATE_LOGGED_FQN ), ImmutableSet.of( dateLogged ) );
+            entityData.put( propertyTypeIdsByFQN.get( STRING_ID_FQN ), ImmutableSet.of( deviceId ) );
+            entityData.put( propertyTypeIdsByFQN.get( FULL_NAME_FQN ), ImmutableSet.of( appPackageName ) );
```

Comment:
Why are you adding on to `entityData`, which already has `FULL_NAME_FQN` and `TITLE_FQN`? Shouldn't you use a new / empty "entity" each time you get an `EntityKey`?

Reply:
There is no need to create a new `Map` each time since the EntityKeys generated only depend on the values mapped by the specified list of propertyIds.